### PR TITLE
[Customer] Repository and registry code cleanup

### PR DESCRIPTION
**Pull request by Den4ik**
### Description (*)
This PR provide code cleanup for repository and registry in customer module

### Fixed Issues (if relevant)

No issues

### Manual testing scenarios (*)

No special scenarios needed

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by rodrigowebjump — CHANGES_REQUESTED** [review]
Hi @Den4ik 

Please review the use of resource models in your change. Check if it is possible to use the Customer Repository methods in these cases. 

Thanks

**Comment by Den4ik** [comment]
@rodrigoobiassi Repository use registry for retrieving customer data from database. I understood your opinion, it's require fully refactoring of registry

File: app/code/Magento/Customer/Model/CustomerRegistry.php [hunk 1]
```diff
@@ -77,8 +92,9 @@ public function retrieve($customerId)
         if (isset($this->customerRegistryById[$customerId])) {
             return $this->customerRegistryById[$customerId];
         }
-        /** @var Customer $customer */
-        $customer = $this->customerFactory->create()->load($customerId);
+        $customer = $this->customerFactory->create();
+        $this->customerResourceFactory->create()
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Model/CustomerRegistry.php:96]: Maybe I'm missing something, but why not using `CustomerRepositoryInterface->getById` there?

Comment (Den4ik) [reply app/code/Magento/Customer/Model/CustomerRegistry.php:96]: @lbajsarowicz Customer repository call registry
```
public function getById($customerId)
    {
        $customerModel = $this->customerRegistry->retrieve($customerId);
        return $customerModel->getDataModel();
    }
```

Comment (lbajsarowicz) [reply app/code/Magento/Customer/Model/CustomerRegistry.php:96]: Aww I see :-) Thanks!

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Customer/Model/CustomerRegistry.php (hunk 1 above)
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/CustomerRegistry.php:96]: Resource model shouldn't contain any data inside, so it seems to me that we need to get customer resource model, not factory of it.
Is there any reason why you used factory there?

File: app/code/Magento/Customer/Model/CustomerRegistry.php [hunk 2]
```diff
@@ -148,9 +164,9 @@ public function retrieveSecureData($customerId)
         if (isset($this->customerSecureRegistryById[$customerId])) {
             return $this->customerSecureRegistryById[$customerId];
         }
-        /** @var Customer $customer */
+        /** @var CustomerResourceFactory $customer */
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/CustomerRegistry.php:167]: It doesn't looks like correct fix, as phpdoc block shows it should return Customer model. Could you double check it?

File: app/code/Magento/Customer/Model/CustomerRegistry.php [hunk 3]
```diff
@@ -48,21 +54,30 @@ class CustomerRegistry
...
         CustomerFactory $customerFactory,
         CustomerSecureFactory $customerSecureFactory,
-        StoreManagerInterface $storeManager
+        StoreManagerInterface $storeManager,
+        CustomerResourceFactory $customerResourceFactory = null
     ) {
         $this->customerFactory = $customerFactory;
         $this->customerSecureFactory = $customerSecureFactory;
         $this->storeManager = $storeManager;
+        $this->customerResourceFactory = $customerResourceFactory
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/CustomerRegistry.php:79]:
```suggestion
        $this->customerResource = $customerResource
```

File: app/code/Magento/Customer/Model/CustomerRegistry.php [hunk 4]
```diff
@@ -48,21 +54,30 @@ class CustomerRegistry
...
+            ?? ObjectManager::getInstance()->get(CustomerResourceFactory::class);
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/CustomerRegistry.php:80]:
```suggestion
        $this->customerResource = $customerResource;
```

File: app/code/Magento/Customer/Model/CustomerRegistry.php [hunk 5]
```diff
@@ -48,21 +54,30 @@ class CustomerRegistry
...
      * @param CustomerSecureFactory $customerSecureFactory
      * @param StoreManagerInterface $storeManager
+     * @param CustomerResourceFactory $customerResourceFactory
      */
     public function __construct(
         CustomerFactory $customerFactory,
         CustomerSecureFactory $customerSecureFactory,
-        StoreManagerInterface $storeManager
+        StoreManagerInterface $storeManager,
+        CustomerResourceFactory $customerResourceFactory = null
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/CustomerRegistry.php:74]: 1. This class isn't marked as `@api`, so as part of 2.4.0 we should add NOT optional arguments.
2. Resource model shouldn't contain any data inside, so there is no need to use factory for this model. 
3. it's better name variable as `$customerResourceModel` to make it more clear

Please adjust it like this:
```suggestion
        CustomerResource $customerResourceModel = null
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @Den4ik,
Will you be able to update your PR?

**Comment by Den4ik** [comment]
Hi @ihor-sviziev Yes, I'll work on update on weekends

**Comment by ihor-sviziev** [comment]
Hi @Den4ik,
I am closing this PR now due to inactivity.
Please reopen it and update with requested changes if you would like to continue.
Thank you for the contribution!

**Pull request closed without merging**